apiserver/facades/client/cloud/backend.go

```diff
@@ -97,17 +101,20 @@ type statePoolShim struct {
 
 // NewModelPoolBackend creates a model pool backend based on state.StatePool.
 func NewModelPoolBackend(st *state.StatePool) ModelPoolBackend {
-	return statePoolShim{st}
+	return statePoolShim{StatePool: st}
 }
 
 // GetModelCallContext implements ModelPoolBackend.GetModelCallContext.
-func (s statePoolShim) GetModelCallContext(modelUUID string) (credentialcommon.PersistentBackend, context.ProviderCallContext, error) {
+// The responsibility of releasing the model state is left to the caller.
+func (s statePoolShim) GetModelCallContext(modelUUID string) (credentialcommon.PersistentBackend, context.ProviderCallContext, ModelCallContextReleaser, error) {
 	modelState, err := s.StatePool.Get(modelUUID)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
-	defer modelState.Release()
-	return credentialcommon.NewPersistentBackend(modelState.State), context.CallContext(modelState.State), err
+	return credentialcommon.NewPersistentBackend(modelState.State),
```

Comment:
Definitely a bit of a thinko to return an object that you just released back to the caller.
Nice catch!